stan/math/prim/mat/fun/value_of_rec.hpp

```diff
@@ -22,12 +22,33 @@ namespace stan {
     template <typename T, int R, int C>
     inline Eigen::Matrix<double, R, C>
     value_of_rec(const Eigen::Matrix<T, R, C>& M) {
-      using stan::math::value_of_rec;
-      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
-      for (int j = 0; j < M.cols(); j++)
-        for (int i = 0; i < M.rows(); i++)
-          Md(i, j) = value_of_rec(M(i, j));
-      return Md;
+      Eigen::Matrix<double, R, C>
+        result(M.rows(), M.cols());
+
+      double * data_r = result.data();
+      const T * data_m = M.data();
+
+      int S = M.size();
+      for (int i=0; i < S; i++)
+        data_r[i] = value_of_rec(data_m[i]);
+      return result;
+    }
```

Comment:
I understand not doing the double indexing --- that's slow.  But I think the bit after the declaration/construction of result, we should just do the following unless you can demonstrate that what you did with all the temporaries is faster.

```
for (int i = 0; i < M.size(); ++i)
  result(i) = value_of(M(i));
```

We know the value can't be recursive through Eigen in any of our code, so we don't need `value_of_rec`.